arch/src/x86_64/mod.rs

```diff
@@ -338,6 +341,149 @@ impl CpuidPatch {
     }
 }
 
+pub struct CpuidFeatureEntry {
+    pub function: u32,
+    pub index: u32,
+    pub feature_reg: CpuidReg,
+}
+
+impl CpuidFeatureEntry {
+    fn predefined_feature_entry_list() -> Vec<CpuidFeatureEntry> {
+        vec![
+            // Leaf 0x1, ECX/EDX, feature bits
+            CpuidFeatureEntry {
+                function: 1,
+                index: 0,
+                feature_reg: CpuidReg::ECX,
+            },
+            CpuidFeatureEntry {
+                function: 1,
+                index: 0,
+                feature_reg: CpuidReg::EDX,
+            },
+            // Leaf 0x7, EAX/EBX/ECX/EDX, extended features
+            CpuidFeatureEntry {
+                function: 7,
+                index: 0,
+                feature_reg: CpuidReg::EAX,
+            },
+            CpuidFeatureEntry {
+                function: 7,
+                index: 0,
+                feature_reg: CpuidReg::EBX,
+            },
+            CpuidFeatureEntry {
+                function: 7,
+                index: 0,
+                feature_reg: CpuidReg::ECX,
+            },
+            CpuidFeatureEntry {
+                function: 7,
+                index: 0,
+                feature_reg: CpuidReg::EDX,
+            },
+            // Leaf 0x7 subleaf 0x1, EBX/ECX/EDX, extended features
+            CpuidFeatureEntry {
+                function: 7,
+                index: 1,
+                feature_reg: CpuidReg::EBX,
+            },
+            CpuidFeatureEntry {
+                function: 7,
+                index: 1,
+                feature_reg: CpuidReg::ECX,
+            },
+            CpuidFeatureEntry {
+                function: 7,
+                index: 1,
+                feature_reg: CpuidReg::EDX,
+            },
+            // Leaf 0x8000_0001, ECX/EDX, CPUID features bits
+            CpuidFeatureEntry {
+                function: 0x8000_0001,
+                index: 0,
+                feature_reg: CpuidReg::ECX,
+            },
+            CpuidFeatureEntry {
+                function: 0x8000_0001,
+                index: 0,
+                feature_reg: CpuidReg::EDX,
+            },
```

Comment:
Aren't we missing SGX (leaf 0x12 if IIRC), and KVM features (`0x40000000` and `0x40000001`) https://www.kernel.org/doc/html/latest/virt/kvm/cpuid.html

Reply:
I added `KVM_CPUID_FEATURES ` bits leaf "0x4000_0001`, EAX/EBX/ECX/EDX.  I am not sure whether we should do the same for `KVM_CPUID_SIGNATURE`, as they are not feature bits and should be checked in a different way (other than checking for feature subset). For example, it looks more reasonable to check that EBX/ECX/EDX should always be `KVMKVMKVM` while the value of `EAX` should be smaller from the source VM. Any thoughts? @sboeuf @rbradford

Reply:
Yes you're right. `0x4000_0000` `EAX` should be smaller or equal on the source. About the other registers it's up to you, either you check the value is `KVMKVMKVM` but this might not work for `mshv`, or you simply check the value is the same between source and destination.

Reply:
Check the value is the same as the source for EAX. Does your code correctly handle the case where the feature leaf is not present at the source (i.e. when the KVM leaf is not present because you're running on MSHV.)

Reply:
We were fine with checking compatibility only for "bitwise subset". To add check on leaf `0x4000_00000` ("KVM_CPUID_SIGNATURE"), I added a new commit with more fine-grained checks for compatibility. We now can check for "equal" and "not greater than".

PTAL.  All changes are in the last (4th) commit (small fixes from last week are squashed).